themes/default/content/blog/csharp-static-analyzer/index.md

```diff
@@ -0,0 +1,132 @@
+
+This however means breaking each and every existing C# program running with Pulumi since the removal of the empty constructor. Here at Pulumi, we want to ensure that changes to our APIs are backward-compatible so this is a no-go.
+
+The problem above looks like the compiler _could_ have detected if it somehow knew which properties were required or and which were optional. In the case of the generated APIs, this requiredness information is only encoded on properties with a special attribute `[Input("propertyName", required: true)]` which unfortunately is Pulumi-specific and the compiler has no way of detecting it and reporting an error or a warning to the developer at compile-time.
+
+## Solution
+
+What do we do? Static code analysis to the rescue! If the compiler can't do it, we can build an extension and have detect these types of errors for the developers.
+
+Enter [PulumiCSharpAnalyzer](https://github.com/Zaid-Ajaj/pulumi-csharp-analyzer): a Roslyn-based code analyzer that adds compile-time checks to your Pulumi code and reports it as part of the build.
+
+Available today as a NuGet package, you can install the analyzer into your project as follows:
+
+```
+dotnet add package PulumiCSharpAnalyzer
+```
+
+Now your IDE, whether that is Rider, Visual Studio or VS Code with OmniSharp should report the error inside the IDE like any compile error or warning. Here is for example how it looks like inside Rider:
+
+![missing property Kind when initializing resource arguments](./missing-kind-property.png)
+
+Not just your IDE, it integrates into your build pipeline too. When you `dotnet build` your project, the detected errors will show up there:
+
+```bash
+> dotnet build
+Microsoft (R) Build Engine version 17.2.0+41abc5629 for .NET
+Copyright (C) Microsoft Corporation. All rights reserved.
+
+  Determining projects to restore...
+  All projects are up-to-date for restore.
+/Users/zaid/projects/AzureExample/Program.cs(11,63): warning MissingRequiredProperty: Missing required property Kind when initializing properties of type StorageAccountArgs [/Users/zaid/projects/AzureExample/AzureExample.csproj]
+  AzureExample -> /Users/zaid/projects/AzureExample/bin/Debug/net6.0/AzureExample.dll
+
+Build succeeded.
+```
+
+Besides detecting missing properties when creating resources, the analyzer also detects missing properties for function invokes which follow a similar pattern:
+
+![missing required parameter when invoking a SDK function](./error-missing-required-properties-function-invokes.png)
+
+Finally, an anti-pattern that the analyzer can detect is creating resources within an `.Apply(...)` call:
+
+![a warning being reported when creating a resource inside apply call](./resource-inside-apply.png)
+
+We always recommend developers to avoid creating resources inside `.Apply(...)` calls because the execution plan of Pulumi will be different during the `pulumi preview` and `pulumi up` phase.
```

Comment:
```suggestion
We always recommend developers avoid creating resources inside `.Apply(...)` calls because the execution plan of Pulumi may be different during the `pulumi preview` and `pulumi up` phases.
```